Specify algorithm in WebSocket JWT verification

shekharnwagh opened the pull request:
## Summary

Improves WebSocket JWT authentication security by explicitly specifying the allowed algorithm during token verification.

## Why

Without explicitly specifying the algorithm, the JWT library accepts whatever algorithm is declared in the token header. This allows attackers to manipulate the algorithm field to bypass signature verification.

By explicitly requiring HS256, we ensure only tokens signed with our expected algorithm are accepted.

shekharnwagh committed:
Specify algorithm in JWT verification

Explicitly require HS256 algorithm when verifying JWT tokens in
the WebSocket authentication flow. This prevents algorithm confusion
attacks where an attacker switches RS256 to HS256 and uses the public
key as an HMAC secret to forge tokens.

shekharnwagh committed:
Remove duplicate assertion in auth test

shekharnwagh committed:
Add tests for JWT algorithm enforcement

chriszarate reviewed (changes requested):
Given this is a security vulnerability that we don't want to repeat, we might want to approach token validation with even more caution. Perhaps we can define wrapper functions for token operations in `utils.ts` that prevent this issue via extra type safety. Something like:

```ts
import jwt, { type Jwt, type JwtPayload, type VerifyOptions } from 'jsonwebtoken';

type WithRequired< T, K extends keyof T > = Omit< T, K > & Required< Pick< T, K > >;

type SignOptionsWithAlgo = WithRequired< VerifyOptions, 'algorithms' >;

export function verifyJwtToken(
	token: string,
	secret: string,
	options: SignOptionsWithAlgo
): JwtPayload | Jwt | string {
	return jwt.verify( token, secret, options );
}
```

Additionally, we could leverage eslint's [`no-restricted-imports` rule](https://eslint.org/docs/latest/rules/no-restricted-imports) to ban importing `jwt` with a pointer to the `utils` file. Then override the rule in that one spot.

shekharnwagh committed:
Add type-safe JWT wrapper to prevent algorithm confusion

Creates verifyJwtToken wrapper in websocket-server/utils that
uses TypeScript to require the algorithms parameter, preventing
algorithm confusion attacks at compile time.

Adds ESLint rule to ban direct jwt.verify imports, forcing use
of the safe wrapper to ensure the 'algorithms' parameter cannot
be accidentally omitted in future code.

shekharnwagh commented:
> Given this is a security vulnerability that we don't want to repeat, we might want to approach token validation with even more caution. Perhaps we can define wrapper functions for token operations in `utils.ts` that prevent this issue via extra type safety. 

Pushed a commit e63e456 to address this. Thank you for the suggestion !

chriszarate commented on websocket-server/auth.ts:
```diff
@@ -36,7 +36,7 @@ function verifyToken( token: string | null | undefined, secret: string ): SyncTo
 	}
 
-	const jwtPayload = jwt.verify( token, secret );
+	const jwtPayload = verifyJwtToken( token, secret, { algorithms: [ 'HS256' ] } ) as unknown;
```
cast is unnecessary 

```suggestion
	const jwtPayload = verifyJwtToken( token, secret, { algorithms: [ 'HS256' ] } );
```

chriszarate reviewed (approved):
nice, thanks!
